File: python/1038.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:
        self.add = 0
        def dfs(node):
            if node:
                dfs(node.right)
                node.val += self.add
                self.add = node.val
                dfs(node.left)
        dfs(root)
        return root

File: java/1038.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int add = 0;
    public TreeNode bstToGst(TreeNode root) {
        dfs(root);
        return root;
    }
    public void dfs(TreeNode node){
        if(node != null){
            dfs(node.right);
            node.val += add;
            add = node.val;
            dfs(node.left);
        }
    }
}